src/pg/mod.rs

```diff
@@ -452,27 +503,53 @@ impl AsyncPgConnection {
     }
 }
 
+/// Collects types that need to be looked up, and causes fake OIDs to be written into the bind collector
+/// so they can be replaced with asynchronously fetched OIDs after the original query is dropped
 struct PgAsyncMetadataLookup {
     unresolved_types: Vec<(Option<String>, String)>,
+    min_fake_oid: u32,
 }
 
 impl PgAsyncMetadataLookup {
-    fn new() -> Self {
+    fn new(bind_collector_0: &RawBytesBindCollector<Pg>) -> Self {
+        let max_hardcoded_oid = bind_collector_0
+            .metadata
+            .iter()
+            .flat_map(|m| [m.oid().unwrap_or(0), m.array_oid().unwrap_or(0)])
+            .max()
+            .unwrap_or(0);
         Self {
             unresolved_types: Vec::new(),
+            min_fake_oid: max_hardcoded_oid + 1,
         }
     }
+
+    fn fake_oids(&self, index: usize) -> (u32, u32) {
+        let oid = self.min_fake_oid + ((index as u32) * 2);
+        (oid, oid + 1)
+    }
 }
 
 impl PgMetadataLookup for PgAsyncMetadataLookup {
     fn lookup_type(&mut self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata {
-        let cache_key =
-            PgMetadataCacheKey::new(schema.map(Cow::Borrowed), Cow::Borrowed(type_name));
-
-        let cache_key = cache_key.into_owned();
+        let index = self.unresolved_types.len();
         self.unresolved_types
             .push((schema.map(ToOwned::to_owned), type_name.to_owned()));
-        PgTypeMetadata::from_result(Err(FailedToLookupTypeError::new(cache_key)))
+        PgTypeMetadata::from_result(Ok(self.fake_oids(index)))
+    }
+}
+
+/// Allows unambiguously determining:
+/// * where OIDs are written in `bind_collector.binds` after being returned by `lookup_type`
+/// * determining the maximum hardcoded OID in `bind_collector.metadata`
+struct SameOidEveryTime {
+    first_byte: u8,
```

Comment:
I likely would just use a 'u32' directly here instead of constructing it everytime in the lookup function.